rust/lance/src/dataset/fragment.rs

```diff
@@ -861,17 +861,32 @@ impl FragmentReader {
             let params = params.clone();
 
             async move {
-                reader
-                    .read_batch(
-                        batch_id as i32,
-                        params,
-                        &projection?,
-                        self.deletion_vec.as_ref().map(|dv| dv.as_ref()),
-                    )
-                    .await
+                // Apply ? inside the task to keep read_tasks a simple iter of futures
+                // for try_join_all
+                let projection = projection?;
```

Comment:
Does `reader` open a file descriptor? If so, should we handle such case during the construction of reader?

Reply:
It does.  We could probably also fix that.  I'll give it a shot.  We need both cases handled though so we will still need this change.

For example, the filter might be `x < 20 AND y < 10`.  From statistics, we might learn that, for batch 7, `x < 20` is always true, and so we don't need to read that column.  Then we would just read column `y` just for that batch.  In this case we do need to open both readers but we need to account for the readers required changing from batch to batch still.

Reply:
I see.